server/src/socketHandlers/socketHandlers.ts

```diff
@@ -184,21 +184,33 @@ export default function initializeSocketHandler(io: Server) {
 
     socket.on("mark_numbers", ({ lobbyId, userId, playerChoice }) => {
       const gameLogic = lobbiesMap[lobbyId].gameLogic;
-      
-      if (gameLogic) {
-        const { row: rowColour, num } = playerChoice;
-        const updatedGameState = gameLogic.makeMove(userId, rowColour, num);
 
-        const responseData = { gameState: updatedGameState };
+      if (gameLogic && gameLogic.hasRolled) {
+        if (playerChoice) {
+          const { row: rowColour, num } = playerChoice;
+          const updatedGameState = gameLogic.makeMove(userId, rowColour, num);
 
-        io.to(lobbyId).emit("update_markedNumbers", responseData);
-        console.log("Updated game state:", updatedGameState);
+          const responseData = { gameState: updatedGameState };
+
+          io.to(lobbyId).emit("update_markedNumbers", responseData);
+          console.log("Updated game state:", updatedGameState);
+        }
+      }
+
+      if (gameLogic?.haveAllPlayersSubmitted()) {
+        gameLogic.resetAllPlayersSubmission();
+        gameLogic.nextTurn();
+
+        const serializedGameState = lobbiesMap[lobbyId].serializedGameLogic;
+
+        io.to(lobbyId).emit("turn_ended", { gameState: serializedGameState });
+        console.log("turn ended", serializedGameState);
```

Comment:
I understand why you've done this as you want to send a specific "turn ended" event back. We don't need to change it at the moment but ideally lines 200-202 should be inside of the makeMove() method inside the game logic class as it is an implementation detail that the socketHandler shouldn't know about. It might sound strange but I wonder if a "turn_ended" event is needed at all because making a move and ending a turn does exactly the same thing in App.tsx in client. They just update the game state and I don't think there's anything special about the event yet.

Reply:
yeah I guess we could do that. Does that mean when the markMove function is called and the playermove is marked and updates the gamestate it will also check that allPlayersSubmitted is true and then end turn? Did it work smoothly for each player choice? Yeah I am not sure if potentially having a turn ended would be useful. Possibly when it comes to more intricate gamelogic and also scoring etc it might be useful to have  a definitive turn end if that makes sense. i am not sure though.

Reply:
Yeah, after marking number, right at the end it will check if all players submitted and then switch active player and turn the flag back to false. Ideally we don't want that logic in socketHandlers but I think that if we are aware of it, then it's simple to change it later.

Reply:
i have updated this.